metals/src/main/scala/scala/meta/internal/metals/callHierarchy/CallHierarchyProvider.scala

```diff
@@ -0,0 +1,619 @@
+import scala.meta.internal.semanticdb.SymbolOccurrence
+import scala.meta.internal.{semanticdb => s}
+
+import scala.meta.internal.parsing.Trees
+import scala.meta.internal.remotels.RemoteLanguageServer
+import scala.meta.internal.semanticdb.Scala._
+
+import scala.meta.io.AbsolutePath
+
+import org.eclipse.lsp4j.CallHierarchyPrepareParams
+import org.eclipse.lsp4j.CallHierarchyIncomingCallsParams
+import org.eclipse.lsp4j.CallHierarchyOutgoingCallsParams
+import org.eclipse.lsp4j.CallHierarchyIncomingCall
+import org.eclipse.lsp4j.CallHierarchyOutgoingCall
+import org.eclipse.lsp4j.CallHierarchyItem
+import org.eclipse.{lsp4j => l}
+import scala.meta.Tree
+import scala.meta.Pat
+import scala.meta.Name
+import com.google.gson.JsonElement
+import scala.meta.Term
+import scala.meta.pc.CancelToken
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+import scala.meta.Defn
+import scala.meta.Member
+
+final case class CallHierarchyProvider(
+    workspace: AbsolutePath,
+    semanticdbs: Semanticdbs,
+    buffers: Buffers,
+    definition: DefinitionProvider,
+    references: ReferenceProvider,
+    icons: Icons,
+    compilers: () => Compilers,
+    remote: RemoteLanguageServer,
+    trees: Trees,
+    buildTargets: BuildTargets,
+) extends SemanticdbFeatureProvider
+    with CallHierarchyHelpers {
+
+  override def reset(): Unit = ()
+
+  override def onDelete(file: AbsolutePath): Unit = ()
+
+  override def onChange(docs: TextDocuments, file: AbsolutePath): Unit = ()
+
+  /**
+   * Data that is preserved between a call hierarchy prepare and incoming calls or outgoing calls request.
+   *
+   * @param symbols The set of symbols concerned by the request.
+   * @param visited Symbols that are already visited to handle recursive cases.
+   */
+  private case class CallHierarchyItemInfo(
+      symbols: Array[String],
+      visited: Array[String],
+      isLocal: Boolean,
+      searchLocal: Boolean,
+  )
+
+  private def buildCallHierarchyItem(
+      source: AbsolutePath,
+      doc: TextDocument,
+      occurence: SymbolOccurrence,
+      range: l.Range,
+      visited: Array[String],
+      token: CancelToken,
+  )(implicit ec: ExecutionContext): Future[Option[CallHierarchyItem]] = {
+    occurence match {
+      case SymbolOccurrence(Some(itemRange), symbol, _) =>
+        compilers()
+          .hover(
+            new HoverExtParams(
+              source.toTextDocumentIdentifier,
+              null,
+              itemRange.toLSP,
+            ),
+            token,
+          )
+          .map(hover =>
+            doc.symbols
+              .find(_.symbol == symbol)
+              .map(info => {
+                val item = new CallHierarchyItem(
+                  info.displayName,
+                  info.kind.toLSP,
+                  source.toURI.toString,
+                  range,
+                  itemRange.toLSP,
+                )
+
+                val symbols = (Set(info.symbol) union (
```

Comment:
Alright same thing here, reading through this we have a ton of deeply nested stuff. Let's try to extract some stuff out to improve on the readability and therefore maintainability.